### PR TITLE
🐛Fix UserTask FormField IDs are not Unique Bug

### LINKED ISSUE #1149: 🐛Copy and Paste UserTasks with FormFields does not Work Correct
#### Describe your issue

The FormField ID of an UserTask should be unique, but while copy and pasting UserTasks with FormFields the ID of the FormField does not get generated new.

#### How can others reproduce the issue?

- create some UserTasks with FormFields
- select one or more and copy them
- paste them to the diagram
- click on a new UserTask 
- look at the PropertyPanel and see that the ID field of a selected FormField is bordered red.

#### Possible solution

Override the copy method of bpmn-js and implement a fix.

#### What's your setup

- BPMN-Studio Version (`4.x.x`):
- OS (`Windows/OS X/Linux` + `version`):
- Node (`node --version`):
- NPM (`npm --version`):
- Docker (`docker version --format '{{.Server.Version}}'`):

> You can remove this section if it's irrelevant to the issue
> (for example for certain questions or feature requests).

#### Issue checklist

Please check the boxes in this list after submitting your Issue:

- [ ] I've checked if this issue already exists
- [ ] I've included all the information that i think is relevant
- [ ] I've added logs and/or screenshots (if applicable)
- [ ] I've mentioned PRs and issues that relate to this one

### PR DESCRIPTION
**Pull request by alexanderkasten**
**Changes:**

1. Add Interfaces for IInternalEvent and IDescriptor
2. Add optional fields to IPropertiesElement
3. Check if a UserTask was Copied Before Pasting it to Rename its formField IDs

**Issues:**

Closes #1149 

PR: #1157 

---

See the [examples](#example) for inspiration.

## How can others test the changes?

> Describe how others can test your changes (you can remove this section for typo fixes etc.)

## PR-Checklist

Please check the boxes in this list after submitting your PR:

- [x] You can merge this PR **right now** (if not, please prefix the title with "WIP: ")
- [x] I've tested **all** changes included in this PR.
- [x] I've also reviewed this PR myself before submitting (e.g. for scrambled letters, typos, etc.).
- [x] I've rebased the `develop` branch with my branch before finishing this PR.
- [x] I've **summarized all changes** in a list above.
- [x] I've mentioned all **PRs, which relate to this one**.
- [x] I've prefixed my Pull Request title is according to [gitmoji guide](https://gitmoji.carloscuesta.me/).

## Example

<details>
<summary>
:warning: Before merging please click here to expand and follow the guide.
</summary>
<br>

Please use `:twisted_rightwards_arrows:` at the beginning of your merge commit title.

Example 1:

<pre>
<code>
:twisted_rightwards_arrows: :bug: Fix Wrong Text Decoration at ...
</code>
</pre>

To get your commit message, just copy the first part of this pull request.

Example 2:
<pre>
<code>
**Changes:**

- Fixes Wrong Text Decoration at ...
- Fixes some typos
- ...

**Issues:**

Closes #NumberOfFixedIssue

PR: #NumberOfThisPR
</code>
</pre>
</details>

## Emojis

<details>
<summary>
Expand for a list of most used Emojis.
</summary>
<br>

Please prefix your commit messages with an Emoji.

Ref: https://gitmoji.carloscuesta.me/

| Description              | Glyphe               | Emoji  |
|--------------------------|----------------------|--------|
| Bugfix                   | `:bug:`              | 🐛     |
| Configuration releated   | `:wrench:`           | 🔧     |
| Cosmetic                 | `:lipstick:`         | 💄     |
| Dependencies Downgrade   | `:arrow_down:`       | ⬇️     |
| Dependencies Upgrade     | `:arrow_up:`         | ⬆️     |
| Formatting               | `:art:`              | 🎨     |
| Improving Performance    | `:zap:`              | ⚡️     |
| Initial commit           | `:tada:`             | 🎉     |
| Linter                   | `:rotating_light:`   | 🚨     |
| Miscellaneous            | `:package:`          | 📦     |
| New Feature              | `:sparkles:`         | ✨     |
| Refactoring Code         | `:recycle:`          | ♻️     |
| Releasing / Version tags | `:bookmark:`         | 🔖     |
| Removing Stuff           | `:fire:`             | 🔥     |
| Tests                    | `:white_check_mark:` | ✅     |
| Work In Progress (WIP)   | `:construction:`     | 🚧     |

</details>

### CODE REVIEW

File: src/modules/design/bpmn-io/bpmn-io.ts [hunk 1]
```diff
@@ -405,6 +414,29 @@ export class BpmnIo {
...
+  }
+
+  private _generateRandomId(): string {
+    let randomId: string = '';
+    const possible: string = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
+
+    const randomIdLength: number = 8;
+    for (let i: number = 0; i < randomIdLength; i++) {
+      randomId += possible.charAt(Math.floor(Math.random() * possible.length));
+    }
```
Comment (SteffenKn) [inline src/modules/design/bpmn-io/bpmn-io.ts]:
```suggestion
    }

```

File: src/modules/design/bpmn-io/bpmn-io.ts [hunk 2]
```diff
@@ -405,6 +414,30 @@ export class BpmnIo {
     return returnPromise;
   }
 
+  private _renameFormFields(event: IInternalEvent): IInternalEvent {
+    const formFields: Array<IModdleElement> = event.descriptor.businessObject.extensionElements.values[0].fields.filter((field: IModdleElement) => {
```
Comment (robpal94) [inline src/modules/design/bpmn-io/bpmn-io.ts]: I would split this up a bit:
```suggestion
    extensionElements: YourTypeHere = event.descriptor.buissnesObject.extensionElements.values
    
    if (extensionElements.length() === 0) {
       //Handle Error here
    }
    
    const formFieldExtensionElements: YourTypeHere = extensionElements[0];
    const formFields: Array<IModdleElement> = formFieldExtensionElements.fields.filter((field: IModdleElement) => {
```

Comment (SteffenKn) [reply src/modules/design/bpmn-io/bpmn-io.ts]: I would just do it like this:
```suggestion
    const allFormFields: Array<IModdleElement> = event.descriptor.businessObject.extensionElements.values[0].fields;

    const filteredFormFields: Array<IModdleElement> = allFormFields.filter((field: IModdleElement) => {
```

Comment (robpal94) [reply src/modules/design/bpmn-io/bpmn-io.ts]: Can we ensure that the retrieved array always has at least one entry? If so then do it like @SteffenKn staid.

Comment (alexanderkasten) [reply src/modules/design/bpmn-io/bpmn-io.ts]: @roblen45 Yes

Comment (alexanderkasten) [reply src/modules/design/bpmn-io/bpmn-io.ts]: @robpal94 Yes

**Pull request merged**